cmd/monitor/main.go

```diff
@@ -0,0 +1,218 @@
+	if v, err := note.NewVerifier(*releasePubKey); err != nil {
+		glog.Exitf("Failed to construct release note verifier: %v", err)
+	} else {
+		releaseVerifiers = note.VerifierList(v)
+	}
+
+	monitor := Monitor{
+		st:               st,
+		stateFile:        *stateFile,
+		releaseVerifiers: releaseVerifiers,
+		handler: func(i uint64, r api.FirmwareRelease) error {
+			glog.Infof("%d: Release version %q available at %s", i, r.Revision, r.SourceURL)
+			return nil
+		},
+	}
+
+	if isNew {
+		// This monitor has no memory of running before, so let's catch up with the log.
+		if err := monitor.From(0); err != nil {
+			glog.Exitf("monitor.From(%d): %v", 0, err)
+		}
+	}
+
+	// We've processed all leaves committed to by the tracker's checkpoint, and now we enter polling mode.
+	ticker := time.NewTicker(*pollInterval)
+	defer ticker.Stop()
+	for {
+		lastHead := st.LatestConsistent.Size
+		if err := st.Update(); err != nil {
+			glog.Exitf("Failed to update checkpoint: %q", err)
+		}
+		if st.LatestConsistent.Size > lastHead {
+			glog.V(1).Infof("Found new checkpoint for tree size %d, fetching new leaves", st.LatestConsistent.Size)
+			if err := monitor.From(lastHead); err != nil {
+				glog.Exitf("monitor.From(%d): %v", lastHead, err)
+			}
+		} else {
+			glog.V(2).Infof("Polling: no new data found; tree size is still %d", st.LatestConsistent.Size)
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			// Go around the loop again.
+		}
+	}
+}
+
+// Monitor verifiably checks inclusion of all leaves in a range, and then passes the
+// parsed FirmwareRelease to a handler.
+type Monitor struct {
+	st               client.LogStateTracker
+	stateFile        string
+	releaseVerifiers note.Verifiers
+	handler          func(uint64, api.FirmwareRelease) error
+}
+
+// From checks the leaves from `start` up to the checkpoint from the state tracker.
+// Upon reaching the end of the leaves, the checkpoint is persisted in the state file.
+func (m *Monitor) From(start uint64) error {
+	fromCP := m.st.LatestConsistent
+	pb, err := client.NewProofBuilder(fromCP, m.st.Hasher.HashChildren, m.st.Fetcher)
+
+	if err != nil {
+		return fmt.Errorf("failed to construct proof builder: %v", err)
+	}
+	for i := start; i < fromCP.Size; i++ {
+		rawLeaf, err := client.GetLeaf(m.st.Fetcher, i)
+		if err != nil {
+			return fmt.Errorf("failed to get leaf at index %d: %v", i, err)
+		}
+		hash := m.st.Hasher.HashLeaf(rawLeaf)
+		ip, err := pb.InclusionProof(i)
+		if err != nil {
+			return fmt.Errorf("failed to get inclusion proof for index %d: %v", i, err)
+		}
+		wantRoot, err := m.st.Verifier.RootFromInclusionProof(int64(i), int64(fromCP.Size), ip, hash)
+		if err != nil {
+			return fmt.Errorf("failed to calculate root from inclusion proof at index %d: %v", i, err)
+		}
+		if !bytes.Equal(wantRoot, fromCP.Hash) {
+			return fmt.Errorf("failed to verify inclusion proof for leaf %d.\ncheckpoint: %x\ncalculated: %x",
+				i, fromCP.Hash, wantRoot)
+		}
```

Comment:
I think there's a `verifier.VerifyInclusionProof` helper which does this build-then-compare for you if you'd like?

Reply:
Cool, switched to use that.